components/dashboard/src/data/featureflag-query.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { Organization, Project, User } from "@gitpod/gitpod-protocol";
+import { useQuery } from "@tanstack/react-query";
+import { getExperimentsClient } from "../experiments/client";
+import { useCurrentProject } from "../projects/project-context";
+import { useCurrentUser } from "../user-context";
+import { useCurrentOrg } from "./organizations/orgs-query";
+
+const featureFlags = {
+    start_with_options: false,
+    showUseLastSuccessfulPrebuild: false,
+    publicApiExperimentalWorkspaceService: false,
+    personalAccessTokensEnabled: false,
+    oidcServiceEnabled: false,
+    // Default to true to enable on gitpod dedicated until ff support is added for dedicated
+    orgGitAuthProviders: true,
+    userGitAuthProviders: false,
+    newSignupFlow: false,
+    linkedinConnectionForOnboarding: false,
+    paymentVerificationFlow: false,
+    team_only_attribution: false,
+};
+
+export const useFeatureFlag = (featureFlag: keyof typeof featureFlags) => {
```

Comment:
If I understand this, instead of getting all of the flags up front, we'd be getting them on-demand. It feels like this interface will push a lot of complexity everywhere we want to consume a flag. They'd all need to handle a flags loading state that they didn't need to before. It was also handy to have a hook that returned all of the flags and you could extract several from it. I'm sure there's a reason for this change, but didn't see a mention of it in the description.

Reply:
The actual change is in this [PR](https://github.com/gitpod-io/gitpod/pull/17428). 

>  instead of getting all of the flags up front

Did we get them upfront? The old code seems to fetch all feature flags one after the other and then setting a single state which would cause a rerender for each. This would happen whenever I switch between projects or orgs. Is that not what happened? Vs. now we only ask for what we need so get at most one rerender per flag and we also cache more.

> They'd all need to handle a flags loading state that they didn't need to before.

You can handle the loading state. We could also return a boolean from `useFeatureFlag`. I haven't used the loading state anywhere. Would you prefer that?

> It was also handy to have a hook that returned all of the flags and you could extract several from it.

This was only used once.